### PR TITLE
Add Profile function

### PR DESCRIPTION
**Pull request by Humbertzhang**
This PR is the rest part of https://github.com/apache/skywalking/issues/5943 following https://github.com/apache/skywalking-python/pull/127 .

It added the profile function into Python Agent, I tested it in normal and concurrent scenario and it works fine.

### CODE REVIEW

File: skywalking/profile/profile_context.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        # if first_span_opname was changed by other plugin, there can start profile as well
+        trace_context.profile_status.update_status(self.attempt_profiling(trace_context,
+                                                                          segment_id,
+                                                                          first_span_opname).get())
+
+    def stop_tracing_profile(self, trace_context: SpanContext):
+        """
+        find tracing context and clear on slot
+        """
+        for idx, profiler in enumerate(self.profiling_segment_slots):
```
[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/profile/profile_context.py:107]: ignore

[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/profile/profile_context.py:107]: ignore

File: skywalking/profile/profile_context.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    def build_snapshot(self) -> Optional[TracingThreadSnapshot]:
+        if not self._profiling_thread.is_alive():
+            return None
+
+        current_time = current_milli_time()
+
+        stack_list = []
+
+        # get thread stack of target thread
+        stack = sys._current_frames().get(int(self._profiling_thread.ident))
```
[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/profile/profile_context.py:198]: ignore

File: skywalking/profile/profile_context.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        for idx, item in enumerate(extracted):
+            if idx > config.profile_dump_max_stack_depth:
+                break
+
+            code_sig = f"{item.filename}.{item.name}: {item.lineno}"
+            stack_list.append(code_sig)
+
+        # if is first dump, check is can start profiling
+        if self.dump_sequence == 0 and not self._profile_context.is_start_profileable():
+            return None
```
[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/profile/profile_context.py:212]: ignore

File: skywalking/profile/profile_context.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/profile/profile_context.py:198]: @sonatype-lift ignore

File: skywalking/profile/profile_service.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            # make stop task schedule and task context
+            current_context = ProfileTaskExecutionContext(task)
+            self.task_execution_context.set(current_context)
+
+            # start profiling this task
+            current_context.start_profiling()
+            logger.debug("profile task [%s] for endpoint [%s] started", task.task_id, task.first_span_op_name)
+
+            millis = task.duration * self.MINUTE_TO_MILLIS
+            self.profile_task_scheduler.schedule(millis, self.stop_current_profile_task, [current_context])
```
Comment (kezhenxu94) [inline skywalking/profile/profile_service.py:143]: We should not always schedule after `task.duration` minutes, right? If the agent restarts, the task will be stopped after `task.duration` minutes after the agent restarted, even at that time the duration may be passed.

We should check whether, for example, `current_time_millis() - task.start_time > task.duration`?

File: skywalking/profile/profile_service.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+            self.remove_from_profile_task_list(need_stop.task)
+
+            # notify profiling task has finished
+            agent.notify_profile_finish(need_stop.task)
+
+    class CheckResult:
+        def __init__(self, success: bool, error_reason: str):
+            self.success = success  # type: bool
+            self.error_reason = error_reason  # type: str
```
Comment (kezhenxu94) [inline skywalking/profile/profile_service.py:163]: Will simply using a tuple `bool, str` make codes more simpler (and Pythonic)? Ignore me if you want to keep it as is

File: skywalking/profile/profile_context.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                    else:
+                        # tell execution context current tracing thread dump failed, stop it
+                        context.stop_tracing_profile(profiler.trace_context)
+
+            need_sleep = (current_loop_start_time + max_sleep_period) - current_milli_time()
+            if not need_sleep > 0:
+                need_sleep = max_sleep_period
+
+            # convert to float second
+            time.sleep(need_sleep/1000)
```
Comment (kezhenxu94) [inline skywalking/profile/profile_context.py:160]:
```suggestion
            time.sleep(need_sleep / 1000)
```

File: skywalking/profile/profile_context.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        if not self.task.first_span_op_name == first_span_opname:
+            return ProfileStatusReference.create_with_none()
+
+        # if out limit started profiling count then stop add profiling
+        if self._total_started_profiling_cnt.get() > self.task.max_sampling_count:
+            return ProfileStatusReference.create_with_none()
+
+        # try to occupy slot
+        if not self._current_profiling_cnt.compare_and_set(using_slot_cnt,
+                                                           using_slot_cnt+1):
```
Comment (kezhenxu94) [inline skywalking/profile/profile_context.py:78]:
```suggestion
                                                           using_slot_cnt + 1):
```

File: skywalking/profile/profile_service.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            self.profile_task_scheduler.schedule(millis, self.stop_current_profile_task, [current_context])
+
+    def stop_current_profile_task(self, need_stop: ProfileTaskExecutionContext):
+        with self._rlock:
+            # need_stop is None or task_execution_context is not need_stop context
+            if need_stop is None or not self.task_execution_context.compare_and_set(need_stop, None):
+                return
+
+            need_stop.stop_profiling()
+            logger.debug("profile task [%s] for endpoint [%s] stoped", need_stop.task.task_id,
```
Comment (kezhenxu94) [inline skywalking/profile/profile_service.py:152]:
```suggestion
            logger.debug("profile task [%s] for endpoint [%s] stopped", need_stop.task.task_id,
```

**Review by kezhenxu94 — APPROVED** [review]
This looks very amazing! I tested it and the result look like below.

Just some minor comments

The tested codes is

```python

def method1():
    time.sleep(0.02)
    return '1'

def method2():
    time.sleep(0.02)
    return method1()

def method3():
    time.sleep(0.02)
    return method2()

if __name__ == '__main__':
    config.service_name = 'provider'
    config.logging_level = 'DEBUG'
    agent.start()

    import socketserver
    from http.server import BaseHTTPRequestHandler

    class SimpleHTTPRequestHandler(BaseHTTPRequestHandler):

        def do_POST(self):
            method3()
            time.sleep(0.5)
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write('{"song": "Despacito", "artist": "Luis Fonsi"}'.encode('ascii'))

    PORT = 9091
    Handler = SimpleHTTPRequestHandler

    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        httpd.serve_forever()
```

Screen Shot 2021-08-22 at 10 46 35

File: skywalking/profile/profile_service.py (hunk 4 above)
Comment (Humbertzhang) [reply skywalking/profile/profile_service.py:143]: I staied at same with Java Agent here, I think the reason for that is to make sure to profile enough time

File: skywalking/profile/profile_service.py (hunk 5 above)
Comment (Humbertzhang) [reply skywalking/profile/profile_service.py:163]: I will modify here

**Review by kezhenxu94 — APPROVED** [review]
Thanks!!! 🎉

**Comment by kezhenxu94** [comment]
@Humbertzhang please consider adding tests for this feature later

**Pull request merged**